ios/RCTHTTPRequestHandler+AuthenticationChallenge.m

```diff
@@ -0,0 +1,189 @@
+}
+
+- (NSSet *)getAllowedDomains {
+  NSDictionary *infoDictionary = [[NSBundle mainBundle] infoDictionary];
+  NSDictionary *appTransportSecurity = [infoDictionary objectForKey:@"NSAppTransportSecurity"];
+  NSDictionary *pinnedDomains = [appTransportSecurity objectForKey:@"NSPinnedDomains"];
+  
+  NSMutableSet *allowedDomains = [NSMutableSet set];
+  
+  [pinnedDomains enumerateKeysAndObjectsUsingBlock:^(NSString *domain, NSDictionary *domainInfo, BOOL *stop) {
+    /* Throw an error if the domain value is an empty string. This validation is important
+     * because the pinned certificate for empty domain has no effect in the ATS and in
+     * addition we don't want to allow a broader domain, quite the opposity.
+     */
+    if (domain.length == 0) {
+      NSString *errorMessage = @"Error: Empty domain found in NSPinnedDomains";
+      NSLog(@"%@", errorMessage);
+      // It intentionally breaks the application
+      // It must be catched in the QA or during development
+      @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                     reason:errorMessage
+                                   userInfo:nil];
+    }
+    
+    /* Throw an error if the domain value is invalid by containing a wildcard prefix.
+     * This is invalid because we already use the "*." prefix internally to make
+     * subdomain validation. One must configure apropriatelly the NSPinnedDomains key
+     * in the Info.plist file.
+     */
+    if ([domain hasPrefix:@"*."]) {
+      NSString *errorMessage = @"Error: Domain with wildcard found in NSPinnedDomains\n\tUse the NSIncludesSubdomains property to allow subdomains";
+      NSLog(@"%@", errorMessage);
+      // It intentionally breaks the application
+      // It must be catched in the QA or during development
+      @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                     reason:errorMessage
+                                   userInfo:nil];
+    }
+    
+    // Check if subdomains are included
+    if ([[domainInfo objectForKey:@"NSIncludesSubdomains"] boolValue]) {
+      // If subdomains are included, add a wildcard entry
+      [allowedDomains addObject:[NSString stringWithFormat:@"*.%@", domain]];
+    } else {
+      [allowedDomains addObject:domain];
+    }
+  }];
+  
+  return [allowedDomains copy];
+}
+
+- (BOOL)checkValidityOfHost:(NSString *)host allowedInDomains:(NSSet *)allowedDomains {
+  for (NSString *allowedDomain in allowedDomains) {
+    if ([allowedDomain hasPrefix:@"*."]) {
+      /* A domain is only valid for the immediate subdomain.
+       * Take for example the "hathor.network" domain, it can only
+       * validate a domain like "mainnet.hathor.network", and must
+       * invalidate a domain like "node1.mainnet.hathor.network".
+       *
+       * This validation only plays by the rules of ATS certification pinning.
+       * See: https://developer.apple.com/news/?id=g9ejcf8y
+       */
```

Comment:
Ok, from what I understand, this implementation enforces domain-level pinning (e.g. mainnet.hathor.network) and strict subdomain validation. Are we certain we want this approach rather than CA certificate pinning? 

With the current implementation:

- We'll need to pin certificates for each domain (or parent domain)
- Certificate rotation would require app updates (this one is a no-go for me)

If we want to pin the CA certificate instead, we'd need to modify this implementation

Reply:
>We'll need to pin certificates for each domain (or parent domain)
I don't see any problem with this right now, specially due to our launch date. We can go with this approach now and improve it later.

>Certificate rotation would require app updates (this one is a no-go for me)
Not necessarily. There are techniques we can use to avoid this.

Reply:
Oh, just to follow this up, I saw that a certificate can have an extension of _alternative hostnames_ or something like that, which can let us to add many other hostnames in just one leaf certificate.

Reply:
> Not necessarily. There are techniques we can use to avoid this.

Which techniques?

Reply:
Oh, you right about "certificate rotation requires app updates". In my answer I was thinking in the scope of a leaf certificate, which by pinning the intermediate certificate we don't need an app update when the leaf certificate changes, but we do need an app update when the intermediate certificate changes, indeed. There is no other technique possible here to avoid app update.